federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,885 @@
+	federationclientset "k8s.io/kubernetes/federation/client/clientset_generated/federation_internalclientset"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	cache "k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/record"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	pkg_runtime "k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"k8s.io/kubernetes/pkg/conversion"
+)
+
+const (
+	// TODO update to 10 mins before merge
+	serviceSyncPeriod = 30 * time.Second
+	clusterSyncPeriod = 100 * time.Second
+
+	// How long to wait before retrying the processing of a service change.
+	// If this changes, the sleep in hack/jenkins/e2e.sh before downing a cluster
+	// should be changed appropriately.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+
+	// client retry count and interval is when accessing a remote kube-apiserver or federation apiserver
+	// how many times should be attempted and how long it should sleep when failure occurs
+	// the retry should be in short time so no exponential backoff
+	clientRetryCount = 5
+
+	retryable = true
+
+	doNotRetry = time.Duration(0)
+
+	UserAgentName = "Federation-Service-Controller"
+	KubeAPIQPS    = 20.0
+	KubeAPIBurst  = 30
+)
+
+type cachedService struct {
+
+	//release_1_3 client returns v1 service
+	lastState *api.Service
+	// The state as successfully applied to the DNS server
+	appliedState *api.Service
+	// cluster endpoint map hold subset info from kubernetes clusters
+	// key clusterName
+	// value is a flag that if there is ready addresses
+	endpointMap map[string]int
+	// cluster service map hold serivice status info from kubernetes clusters
+	// key clusterName
+
+	serviceStatusMap map[string]api.LoadBalancerStatus
+
+	// Ensures only one goroutine can operate on this service at any given time.
+	rwlock sync.Mutex
+
+	// Controls error back-off for procceeding federation service to k8s clusters
+	lastRetryDelay time.Duration
+	// Controls error back-off for updating federation service back to federation apiserver
+	lastFedUpdateDelay time.Duration
+	// Controls error back-off for dns record update
+	lastDNSUpdateDelay time.Duration
+}
+
+type serviceCache struct {
+	rwlock sync.Mutex // protects serviceMap
+	// federation service map contains all service received from federation apiserver
+	// key serviceName
+	fedServiceMap map[string]*cachedService
+}
+
+type ServiceController struct {
+	dns              dnsprovider.Interface
+	federationClient federationclientset.Interface
+	// each federation should be configured with a single zone (e.g. "mycompany.com")
```

Comment:
I did not handle any logic related to zones, this part relies on Quinton. removed comment.